### ISSUE TITLE
Bulk delete reports

### ISSUE THREAD

**Issue by abbyad** [medic/medic-webapp#1000]
During testing and training, numerous messages and reports are submitted that need to be deleted once the instance goes to production, especially since it affects usability and analytics.

Currently the only way for a project manager to delete messages/reports is to go through them individually, which is extremely time consuming. We should look at options for bulk delete first, and perhaps consider a "test mode" for the webapp.

Note that some CHW training may overlap with production use in other districts, so it is not sufficient to delete all messages/reports.

**Comment by mandric** [comment]
We might also consider bulk updates, in the past you could delete/validate/invalidate/modify facility for a selected group of records.

**Comment by abbyad** [comment]
This feature request continues to be requested by many users, including our own PMs and tech leads.

Here is a design that should allow for delete, as well as modify facility and validity:

![image](https://cloud.githubusercontent.com/assets/789512/9694527/75acfe6e-5327-11e5-975f-77a6ca2a5461.png)

This should be compatible with the desktop mode as well. The only significant difference would be the Edit page, which could be either a modal or in the right pane. If it is in the right pane the number of reports selected could be modified before saving the edits. This could be a nice touch, but could lead to confusion as well. It is likely simpler to keep the Edit in a modal for now.

**Comment by bishwas-medic** [comment]
Just been inquired about this feature by one of our project officer. The overlapping schedules for training and reporting for different areas and the slow unreliable internet makes it very very hard for them to delete the training reports and monitor the real reports. Would love to see this feature present in the next release.

**Comment by mandric** [comment]
@bishwas-medic this is next on my list after https://github.com/medic/medic-webapp/issues/1764. If you have any input on usability or UX let us know.  As soon as I have something in the works I will post here for feedback.

**Comment by bishwas-medic** [comment]
@mandric beside the UI that Marc has suggested on one of the comments above, it would be great if we could also have an option for "Select All" feature. That would help in deleting large number of reports in very short time. The feature should be:
* Able to select all reports that fits the filter that is applied.
* Usable only after any filter is applied. Shouldn't be usable when no filter is applied to the reports displayed.
* User restricted. Only admin should be able to select all an delete.
* Have "Deselect All" option.

**Comment by diannakane** [comment]
Does deleting a report delete the schedules created by that report? 

And if a person is created in the test process, does a separate deletion process that needs to happen to remove test data?

**Comment by diannakane** [comment]
Include a stop gap "are you sure?" warning dialogue box before bulk deleting, listing how many records are about to be deleted.

**Comment by diannakane** [comment]
@Lesterng related convo from Slack:

henok While discussing the bulk delete feature, something we were talking about last week was to have the option of editing the reports and assigning a different facility using those same check boxes that will be used for bulk delete.  In other words, you bulk select, then use the edit feature we already have iteratively on each of the selected reports.  
Use Case: When a CHW moves or stops being a part of the system, we need to assign another CHW in her place that will get reminder messages for each of her reports(edited)
[2:13 AM] 
marc There are a lot of high priority items for 0.6, and this is likely a lower priority feature so would likely be pushed off of the initial 0.6 release unless there is a big need for it (the generalized tag feature)

**Comment by Lesterng** [comment]
@diannakane and I talked about Bulk Delete being a web app function and view. CHWs will not need to bulk delete from Reports (History). This will only be a web app function.

**Comment by abbyad** [comment]
This **draft** mockup is pending further design to see how we can better serve our user stories, and intended to clarify the [mockup above](https://github.com/medic/medic-webapp/issues/1000#issuecomment-137854871) for "delete only". 

![image](https://cloud.githubusercontent.com/assets/789512/12684085/ce5c916a-c68a-11e5-8221-da5b3aaa699a.png)

Notes: 
- No “select all” option included since it can select items not seen due to scroll loading. Further design can be done on this to find a way to include it.
- Confirmation dialog mentioning how many items will be deleted
- Remove the checkmark for “validate”, instead use the word as a tag on item. Exact location in list and profile to be determined by design team.

**Comment by browndav** [comment]
@diannakane, @Lesterng, @abbyad: We've tentatively scheduled this issue for the current iteration, right behind #1764 – which means it'll probably end up reaching the front of the queue on February 15th. Is this compatible with the design timeline here?

**Comment by diannakane** [comment]
I shared the mockups with Marc last week and we are reviewing tomorrow. Pending his feedback, they will be ready for implementation as soon as tomorrow.

**Comment by diannakane** [comment]
Latest mockup of Bulk Delete. https://projects.invisionapp.com/share/XU645W0Z5

Will now work on user stories for bulk edit to see if any changes need to be made to accommodate that.

**Comment by garethbowen** [comment]
My feedback on the current mockup...

1. The select button. The coloured bar (aka filter bar) doesn't have any other actions, and the button is styled the same as the filters, so it confuses the design somewhat. Furthermore there actually isn't very much room left in the filters bar on the reports page on smaller desktop screens, and there is absolutely none on mobile. If we find somewhere else to put this we can get the feature on mobile for free. The best place in the current UI for actions relating to the list is in the floating action bar at the bottom, because this is where other actions are currently. The "select all/deselect all" toggle should replace the "select" button in the action bar much like it does now in the filter bar.
2. The delete button. This should definitely be in the floating action bar. Removing this and the select button (1) means we will have enough room to leave the filters in place which would be very useful for filtering the reports and then selecting all.
3. Does select all only select the current page? eg: if there are 100 reports and we're showing the first 50, will clicking "select all" then "delete" remove the 50 shown or all 100?

@diannakane Thoughts?

**Comment by diannakane** [comment]
Thanks for the feedback, Gareth. 

1. Our original design had the select button in the floating action bar, but we realized that completing "select all / deselect all" actions from there was inconsistent with how we are using the floating bars across the app (they open a list of forms, not a center for other actions as we had it -- see old mockup: https://projects.invisionapp.com/share/X65T977SV#/screens/131466288). Once making that change to having the "select" button enter us into "select mode," it felt more appropriate for it to be at the top (modeled after Google Inbox -- screenshot attached). Your comment about the button being confused with filters and the space in this bar on small screens is valid. @Lesterng any ideas? 

2. The filters are disabled when you go into "select mode" so as to avoid confusion related to changing filters in the middle of a selection process (also the way Inbox handles this). We also copied Inbox who has their delete button at the top, and will add to that with an edit button for bulk edit soon. 

3. The expectation is that "select all" will select all records, even those unseen. We know there may be some tech complexity here due to how records load so please advise on what is possible. From a user standpoint, we don't want them to have to redo the delete action for every "batch" shown on the screen. We have implemented a count of selected items at the top and a warning box before deleting to avoid mistakes. 

@abbyad feel free to chime in if I missed anything or if you have suggestions for changes. 

![inbox select mode](https://cloud.githubusercontent.com/assets/1675870/13621274/30c01bfe-e54b-11e5-9fd9-02a690fe3ee5.jpg)

**Comment by estellecomment** [comment]
Nice! A couple thoughts. 

 - Grey out the Delete button when no reports are selected.

 - Maybe this has been discussed already, but is there a use case for filters in the delete process? i.e. delete all reports from this place or this user? I saw the point earlier about avoiding confusion. Filters in their current form will show vs hide the reports, whereas in this mode we would want to select vs unselect. 
That may be out of scope of this issue.

 - I'm a bit weirded out by the top bar (with the logo and tabs) disappearing when you get into delete mode.

**Comment by diannakane** [comment]
Thanks @estellecomment

**Comment by diannakane** [comment]
@garethbowen and @estellecomment:

I had a chat with Lester today. We agree with Gareth that the select button should look functionally different than the filters, but think that's more of an issue with how filters look than how the select button looks. Our quick answer is: "select" won't be shown to CHPs in mobile view right now (it's an admin task that will be completed on web app), so a crowded mobile filter bar isn't a bigger issue than it already is. We're rethinking filter so hopefully that whole bar will be improved by the time we need to bring "select" into mobile. 

+1 to @estellecomment's comment (!) on greying out the delete icon when no reports are selected. Do you need the mockup updated or can we just make a note to implement it with that change?

As for the change to the whole top bar in 'select' mode, we hear you and we're open to suggestions. There's benefit to limiting what people can do and see at a time, similar to how forms are completed in the mobile app, but realize that we're extending a mobile experience into the web app. We received guidance from @abbyad on this design and positive feedback from teammates so my inclination would be to move forward with it and see if anyone complains. Bulk delete needs to be a thoughtful action so I think there are benefits to going into "the zone" and not getting distracted by clicking other tabs.

**Comment by abbyad** [comment]
It is nice to see the conversation on this! 

From my perspective, even if the filter buttons will change it is important that it works with our current design since it will likely be in use long before the filters change. There is nothing stopping an admin user from using this functionality on mobile/tablet view, so it is important to consider the space. 
We can definitely update this once the filters do change.

We previously had the hotdogs on the filter bar as well for additional options (eg Export on Reports tab). We since moved tab specific functions to the action bar, in part to avoid having double hotdogs. It would be consistent with the rest of the app to have the Select button in the left pane's Floating Action Bar. Otherwise, having it in a restored hotdogs would still be preferable to me over a button on the left side of the filter bar.

As for the delete button, I do agree that with our current design it makes sense to have it in the Floating Action Bar in the right pane, but that doesnt work for mobile. Can we have it both in the top bar and in the right pane's action bar?

**Comment by diannakane** [comment]
OK, Lester and I will revisit this. Would one of the devs ( @garethbowen ? ) be available to collaborate on this early next week?

**Comment by garethbowen** [comment]
> ... it makes sense to have it in the Floating Action Bar in the right pane, but that doesnt work for mobile.

I was thinking it'd be in the LHS action bar seeing as it relates to the list. This would work on mobile and desktop leaving the RHS as a readonly view of selected items.

I'm free to collaborate, perhaps straight after the iteration meeting tomorrow?

**Comment by alxndrsn** [comment]
What's the deal with this issue?  Is it still waiting for design work, or can it be picked by a developer?

**Comment by garethbowen** [comment]
It's waiting for design work.

**Comment by diannakane** [comment]
@marc and I are in agreement that we move forward with bulk delete using option B for filter behavior when selecting items. Full set of options are in #design channel; here is how option B works. 

> You are in select mode. You have narrowed your filters down to one catchment area and hit "select all" which is 125 records. Then, you realize that you need to narrow down further to the day of the training. You select the date filter and your current selection is cleared, with 75 unselected records there. You hit "select all" to select these records. 
> 
> You realize that training was over two days, so you add a day to the date filter. Your previous selection is cleared and 25 new records appear in your list. You hit "select all" and your selected count is 100.

This is the option that is currently mocked up in Invision, which removes your selection when you change filters.  It’s a compromise between the simplest option (D, of having to go back to change filters, which effectively clears your selection as well), while increasing transparency of what you’re looking at in the selectable list (without filters in select mode, a user may forget which filters are in place on the list they are working with).

Note that in web the right pane serves as a “stack” of selected messages, and you can view one of them at a time, the way Gmail's Inbox functions. Any highlighted item in either pane will “match” to itself on the other side. This works for option B since you can only select items that you can see (in C for instance, you could have a selected item on the right without a “match” on the left).  This also means that to view a record in select mode, you have to select that item. We demonstrated that in one of the older mockups. 

Mobile: https://projects.invisionapp.com/share/F26YOW2CP#/screens/149585517
Web: https://app.moqups.com/joshnesbit/efFun0NVwj/view

I’d like to alter the warning box language slightly. I’m trying to think of language that won’t make it sound like the items may be deleted or removed.  “Changing the filters will deselect all. Do you want to proceed?” OK / Cancel / Don’t show me this warning again.

For reference:
Principles to design by:
— It’s OK to refresh your work when you switch between screens, rather than maintain multiple work-in-progress screens. We are already doing this when you move to a new tab. 
— Trade powerful features for simpler ones to achieve an easy-to-learn user experience.
— Be as transparent as possible about where you are and what you’re doing in the app. No guessing needed.

**Comment by diannakane** [comment]
In History tab:
-- Add a "select" button to the far left on the FAB.  
-- When Select is clicked, you enter "select mode." An unchecked checkbox appears next to every record. The FAB changes to show: Select All / Delete / Close (mobile) and Select All / Close (web). The right FAB (web) changes to "verify all / edit all / delete all" with "verify all" and "edit all" currently disabled.  
-- Filter choices made prior to entering select mode are still enabled. Likewise, any filters chosen in select mode persist after returning to normal mode. 
-- Filter choices made in select mode, after any item has already been selected, incur a warning: "Changing the filters will deselect all. Do you want to proceed?” >> YES / Cancel If no items have been selected, filters can be adjusted without warning.
-- When any checkbox is selected, a count appears at the top of the records list (mobile) and the top of the right pane (web). 
-- Web only: Selected items are listed in the right pane in the same format as on the left, with the addition of a "x" on the record which can remove it from the list. When the record is clicked in the right pane, it opens to reveal the full form contents (see Google Inbox example below). Clicking again closes the contents. 
-- Web only: An item may be removed from the selected list by unchecking it in the left pane or clicking the x on the record in the right pane. Clicking the x gives you a warning: "Do you want to remove this record from the selected list? This action does not delete the record. Do you want to proceed?" >> Yes / Cancel
-- Clicking close exits select mode and returns you to the "normal" FAB.
-- When clicking delete, warning dialogue: "You are about to delete xx records. Do you want to proceed?" >> Yes / Cancel
**-- Bulk Delete function is restricted to managers (admin?) access only. ** "Select" mode can be hidden from CHW users, or if easier, can appear but be disabled.

MOBILE: https://projects.invisionapp.com/share/F26YOW2CP#/screens/149585517

WEB: 
![web_normal mode](https://cloud.githubusercontent.com/assets/1675870/14829783/20ac5796-0ba2-11e6-8307-5e79b49bc422.png)
![web_select mode](https://cloud.githubusercontent.com/assets/1675870/14829784/229f26aa-0ba2-11e6-93b4-2a09bf3eb9ef.png)

Web right pane record stack example: 
![closed](https://cloud.githubusercontent.com/assets/1675870/14829551/e37d5b1e-0ba0-11e6-968a-7960d9705c2b.png)
Web right pane open / view record example: 
![open](https://cloud.githubusercontent.com/assets/1675870/14829557/e7a8a9aa-0ba0-11e6-9806-d83850b49ae1.png)

**Comment by abbyad** [comment]
Thanks @diannakane, there is plenty there to start with. Assigning to @garethbowen to implement or assign as needed. Also, given the length of this issue, feel free to open a dev issue if you prefer.

**Comment by alxndrsn** [comment]
@diannakane don't forget to cancel your amazon prime subscription at the end of the month!

**Comment by estellecomment** [comment]
Cool, good job!

A couple thoughts (apologies if I should have come up with these details earlier and now's not a good time, in which case ignore it)

1. Can we have different icons for Verify all, Edit all, Delete all? To make it more obvious the action is different. And maybe change the color of the FAB bar in Select Mode? (maybe the same blue as the checkboxes, to follow the color scheme)

2. Can the "get out of select mode" button (which is "Close" with "X" icon right now) be the opposite of the "get into select mode" button (which is "Select" with checkbox icon right now)? Have a single button that stays in the same place and toggles select mode?
(note that in select mode, you have the "Select all" button which will have the same checkbox icon. So maybe "Select mode"'s icon could be a multiple checkbox picture, or something?)

(I can make mockups of 1. and 2. if that helps) 

3. I'm not sure I get the point of using the right pane's "X"s to unselect docs. It seems complicated to select/unselect in two different places, and using two different visuals (checkbox vs "X"). The modal ("this will not delete the record") does help to understand what's going on, but the fact we need a modal to explain is a bad sign... 
I'm assuming the issue is that the left and right pane end up scrolled separately, so the selected record is visible on one pane but not the other, so you need to unselect in both panes? 
If yes, then we can resync the two panes' scrolling. What if when you click a selected doc in the right pane, it scrolls the left-pane to the corresponding selected element and highlights it? (and vice versa) Then you can do all your unselecting in the left pane and don't need the "X".

**Comment by estellecomment** [comment]
Here goes mockups, couldn't resist, excuse the extreme hideousness.

Select mode off : 
![selectmode2](https://cloud.githubusercontent.com/assets/911434/14845380/17607f48-0c5e-11e6-8add-fe0bd7cfcb1c.png)

Select mode on : 
![selectmode2_on](https://cloud.githubusercontent.com/assets/911434/14845384/1b8eeff0-0c5e-11e6-870c-aebb074111c9.png)

(I left Export in select mode, in case we want to export selection, some day)

**Comment by SCdF** [comment]
> @diannakane don't forget to cancel your amazon prime subscription at the end of the month!

Gotta watch dat Black Sails tho…

**Comment by diannakane** [comment]
Thanks @estellecomment for the thoughtful suggestions. We need to close the design phase for this issue, and I'm happy to explain why we settled on the designs we did for points (1) and (2) above. Gareth and Marc were our dev resources and were heavily involved in those.  To clarify, the icon for "select mode" isn't actually blue -- just a bad mockup! I expect the styling to match the hi-fi mockups presented for mobile in Invision. 

To your last point, it's interesting that you say that because the interaction you describe is actually how I originally mocked it up. After talking to Lester, we added the "x" to remove the record from the right at the last minute to simplify the request to the devs (though admittedly, without talking to a dev). If it's not too much code to implement the syncing of the right and left pane so a record selected on one side will scroll to "match" the record on the other side (and hence have access to the checkbox to remove the record from the left only), that works for me. @garethbowen what do you think?

**Comment by garethbowen** [comment]
Scrolling between is one option. I was looking at changing the icon to a "minus" or similar. As always, I'll play around with the design during development and see what I end up with.

**Comment by garethbowen** [comment]
Some design changes summarised from a Slack conversation...

1. This won't be restricted to admins. If a user has permission to delete a record they will also be able to bulk delete that record. In practice though, LG CHPs can't do either.
2. Select All should select all records that match the filter whether or not they are rendered in the list (ie: outside of the current page). All of these should be able to be reviewed on the RHS either through pagination or enough performance work that they can all be shown at once.
3. Given (2) it's difficult to sync scrolling as items on the RHS may not be shown on the LHS. Look at ways to deselect items on the RHS.
4. Redesign the expanded view to have the same header as the collapsed view so it's not so jarring when expanding/collapsing the record (mockup below). Use this design on the single view mode also to be consistent.

![multi-select](https://cloud.githubusercontent.com/assets/2064938/14999746/4d9c3026-11df-11e6-9a2c-8d7292c545cc.png)

**Comment by diannakane** [comment]
Thanks, Gareth. One quick point of clarification:

   1. This won't be restricted to admins. If a user has permission to
   delete a record they will also be able to bulk delete that record. In
   practice though, LG CHPs can't do either.

We will be able to attribute bulk delete as a unique feature to admins
while still giving users the ability to single delete. In the case of our
current projects, CHWs can't single delete, so not particularly relevant at
this point in time.

On Tue, May 3, 2016 at 3:02 PM, Gareth Bowen <notifications@github.com>
wrote:

> Some design changes summarised from a Slack conversation...
>
>    1. This won't be restricted to admins. If a user has permission to
>    delete a record they will also be able to bulk delete that record. In
>    practice though, LG CHPs can't do either.
>    2. Select All should select all records that match the filter whether
>    or not they are rendered in the list (ie: outside of the current page). All
>    of these should be able to be reviewed on the RHS either through pagination
>    or enough performance work that they can all be shown at once.
>    3. Given (2) it's difficult to sync scrolling as items on the RHS may
>    not be shown on the LHS. Look at ways to deselect items on the RHS.
>    4. Redesign the expanded view to have the same header as the collapsed
>    view so it's not so jarring when expanding/collapsing the record (mockup
>    below). Use this design on the single view mode also to be consistent.
>
> [image: multi-select]
> <https://cloud.githubusercontent.com/assets/2064938/14999746/4d9c3026-11df-11e6-9a2c-8d7292c545cc.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/medic/medic-webapp/issues/1000#issuecomment-216678306>
>



-- 
P.S. Join our community. Sign up for Medic Mobile's newsletter
<http://medicmobile.us8.list-manage.com/track/click?u=e5b07d2afb4b273eaff8e3c53&id=a3129fc799&e=c511575255>
.


Dianna Kane | Senior Designer
Medic Mobile <http://www.medicmobile.org/> +1 216 262 2204
@*M*edic <http://www.twitter.com/medic> Facebook
<https://www.facebook.com/MedicMobile> @diannakane

**Comment by estellecomment** [comment]
About 2. : if select-all selects records that are too far down the list in the LHS, then they are also too far down the list in the RHS, since we can't display it all in RHS either. So the issue is syncing the scrolls.

I think the problems are coming from having two panels. The information we show is redundant, it's shown twice.
Can we merge the two panels, or get rid of one? Can't we do everything in the RHS google-inbox-style panel? Something that would look like this : 

![229f26aa-0ba2-11e6-93b4-2a09bf3eb9ef](https://cloud.githubusercontent.com/assets/911434/15009282/d1624408-11e5-11e6-960b-d0c2dd457e5c.png)

(desktop would be basically similar, if not the same, to mobile.)

**Comment by estellecomment** [comment]
Sorry just seeing now that it's been discussed in the slack channel. It was dropped too fast though!! I think single pane would be great.

**Comment by diannakane** [comment]
One vs. Two panels is a design topic much larger than this issue, but it's
good to gather insights into where we run into challenges with our current
container.

On Wed, May 4, 2016 at 2:11 AM, Estelle Comment <notifications@github.com>
wrote:

> Sorry just seeing now that it's been discussed in the slack channel. It
> was dropped too fast though!! I think single pane would be great.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/medic/medic-webapp/issues/1000#issuecomment-216804340>
>



-- 
P.S. Join our community. Sign up for Medic Mobile's newsletter
<http://medicmobile.us8.list-manage.com/track/click?u=e5b07d2afb4b273eaff8e3c53&id=a3129fc799&e=c511575255>
.


Dianna Kane | Senior Designer
Medic Mobile <http://www.medicmobile.org/> +1 216 262 2204
@*M*edic <http://www.twitter.com/medic> Facebook
<https://www.facebook.com/MedicMobile> @diannakane

**Comment by estellecomment** [comment]
@garethbowen can you give me a summary of the code change? How stuff works, which file to start reading at, stuff like that?

**Comment by estellecomment** [comment]
also, for the next time, I'm happy reviewing in pieces. Like if you do the backend first, review that, then the front-end piece, then the bells and whistles (or whatever division makes sense)

**Comment by garethbowen** [comment]
I guess I'd start in the Reports controller. The crux of the change is that selected is now an array instead of just the selected object. This applies to both select mode and view mode to keep the same code being used in both. If you can follow this and the associated template changes then hopefully everything else will flow from there.

One change I thought you might be interested in is I'm now passing `model` in to the confirm controller. I'm not sure if this works for you, but it means I can say "are you sure you want to delete 1,605,302 records?" in the dialog.

**Comment by estellecomment** [comment]
The checkboxes don't stay checked, they uncheck in a seemingly random way. The selection seems to be remembered right though.
![screen shot 2016-05-24 at 16 35 47](https://cloud.githubusercontent.com/assets/911434/15507927/c956c58e-21cd-11e6-8ed3-c53126caa45d.png)

**Comment by estellecomment** [comment]
Select all is either slow or doesn't work, I get a spinner.

**Comment by estellecomment** [comment]
Select all, then unselect all, then select all gives me these console errors : 
```
angular.js:13550 TypeError: Cannot read property 'rows' of undefined
    at getIntersection (search.js:53)
    at search.js:73
    at processQueue (angular.js:15961)
    at angular.js:15977
    at Scope.$eval (angular.js:17229)
    at Scope.$digest (angular.js:17045)
    at Scope.$apply (angular.js:17337)
    at HTMLAnchorElement.<anonymous> (angular.js:25023)
    at HTMLAnchorElement.jQuery.event.dispatch (jquery.js:4435)
    at HTMLAnchorElement.elemData.handle (jquery.js:4121)(anonymous function) @ angular.js:13550(anonymous function) @ angular.js:10225processQueue @ angular.js:15969(anonymous function) @ angular.js:15977Scope.$eval @ angular.js:17229Scope.$digest @ angular.js:17045Scope.$apply @ angular.js:17337(anonymous function) @ angular.js:25023jQuery.event.dispatch @ jquery.js:4435elemData.handle @ jquery.js:4121
2016-05-24 16:41:21.161 angular.js:13550 TypeError: Cannot read property 'rows' of undefined
    at getIntersection (search.js:53)
    at search.js:73
    at processQueue (angular.js:15961)
    at angular.js:15977
    at Scope.$eval (angular.js:17229)
    at Scope.$digest (angular.js:17045)
    at Scope.$apply (angular.js:17337)
    at HTMLAnchorElement.<anonymous> (angular.js:25023)
    at HTMLAnchorElement.jQuery.event.dispatch (jquery.js:4435)
    at HTMLAnchorElement.elemData.handle (jquery.js:4121)
```

**Comment by estellecomment** [comment]
- [ ] When I get into select mode, the green dots shift to the left and the text shifts to the right, it looks weird (which you probably saw but tried to ignore because CSS is annoying)(yay code review!)

**Comment by estellecomment** [comment]
- [ ] To get out of select mode, the button says 'Close', but you're not really closing anything, you're changing modes. I vote for 'Done' instead

**Comment by estellecomment** [comment]
- [ ] If you select 2 records, then unselect 1 from the 'X' in the RHS, then 'Delete All', it says 'Are you sure you want to delete 2 records'

**Comment by estellecomment** [comment]
- [ ] Select something in filter, then get in select mode. Filters stay enabled.

**Comment by garethbowen** [comment]
> The checkboxes don't stay checked...

If Changes fired while in select mode it got out of sync. Changes will fire if you're selecting a doc you haven't read before (for example). This is now fixed.

> If you select 2 records, then unselect 1 from the 'X' in the RHS, then 'Delete All', it says 'Are you sure you want to delete 2 records'

Fixed.

> Select something in filter, then get in select mode. Filters stay enabled.

This is by design. You just can't change filters once you've selected something in select mode.

> To get out of select mode, the button says 'Close', but you're not really closing anything, you're changing modes. I vote for 'Done' instead

This is as in the mockup. Let's leave it for UAT. It's translated so it can be changed in the field anyway.

> When I get into select mode, the green dots shift to the left and the text shifts to the right, it looks weird

Yeah... the problem here is different browsers render checkboxes differently. Mind if we leave it for UAT and see how much it annoys people?

> Select all is either slow or doesn't work, I get a spinner.

It's very slow. I've split out #2302 to make this faster but didn't want to derail this issue. I've now completed #2302 and hope to merge it into this branch sometime today.

> Select all, then unselect all, then select all gives me these console errors

Interesting... I'll look into it.

**Comment by garethbowen** [comment]
> Select all, then unselect all, then select all gives me these console errors

Search service has been refactored and I can't reproduce this now.

I've merged in the latest from develop including the Search performance improvements so Select All should perform much better now.

@estellecomment This is now ready for review again. :)

**Comment by estellecomment** [comment]
> the green dots shift to the left and the text shifts to the right [...] Mind if we leave it for UAT and see how much it annoys people?

I do think it should be fixed forreal. If UAT doesn't complain, I will re-complain :) Lemme know if you need help. I'm seeing it on Chrome 50.0.2661.102, are you using Firefox?

> This is by design. You just can't change filters once you've selected something in select mode.

Oh I see. I expected it would be the switching to select mode that would block the filters, rather than the first selection. 
It would be functionally equivalent (no changing filters in selection process), have we discussed that option? 
I can't figure out which of the two would be more usable. 
On the one hand it's easier to understand : "select mode => no filter, no select mode => filter". 
But maybe it's more annoying because you need to get back out of select mode to fiddle filters. Though you're losing your existing selection either way, really.
Thoughts @diannakane ?

**Comment by diannakane** [comment]
@estellecomment Yeah, they're functionally equivalent, but this way
provides you with the ability to start with a pre-filtered list and adjust
your filters without the "back" and "re-enter select mode" steps. Here's a
summary of our conclusions. Full conversation in the #design channel,
around April 19th.

Update on filter behavior discussion: After careful consideration and
reflection with various teammates, here are a few design principles that
I’d like to add to our app:

— It’s OK to refresh your work when you switch between screens, rather than
maintain multiple work-in-progress screens. We are already doing this when
you move to a new tab.

— Trade powerful features for simpler ones to achieve an easy-to-learn user
experience.

— Be as transparent as possible about where you are and what you’re doing
in the app. No guessing needed.

With these in mind, @marc <https://medic.slack.com/team/marc> and I are in
agreement that we move forward with option B for filter behavior when
selecting items.  This is the option that is currently mocked up in
Invision, which removes your selection when you change filters.  It’s a
compromise between the simplest option (D, of having to go back to change
filters, which effectively clears your selection as well), while increasing
transparency of what you’re looking at in the selectable list (without
filters in select mode, a user may forget which filters are in place on the
list they are working with).
I’d like to alter the warning box language slightly. I’m trying to think of
language that won’t make it sound like the items may be deleted or removed.
 “Changing the filters will deselect all. Do you want to proceed?” OK /
Cancel / Don’t show me this warning again.

On Wed, May 25, 2016 at 6:16 AM, Estelle Comment <notifications@github.com>
wrote:

> the green dots shift to the left and the text shifts to the right [...]
> Mind if we leave it for UAT and see how much it annoys people?
>
> I do think it should be fixed forreal. If UAT doesn't complain, I will
> re-complain :) Lemme know if you need help. I'm seeing it on Chrome
> 50.0.2661.102, are you using Firefox?
>
> This is by design. You just can't change filters once you've selected
> something in select mode.
>
> Oh I see. I expected it would be the switching to select mode that would
> block the filters, rather than the first selection.
> It would be functionally equivalent (no changing filters in selection
> process), have we discussed that option?
> I can't figure out which of the two would be more usable.
> On the one hand it's easier to understand : "select mode => no filter, no
> select mode => filter".
> But maybe it's more annoying because you need to get back out of select
> mode to fiddle filters. Though you're losing your existing selection either
> way, really.
> Thoughts @diannakane <https://github.com/diannakane> ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/medic/medic-webapp/issues/1000#issuecomment-221572408>
>



-- 
P.S. Join our community. Sign up for Medic Mobile's newsletter
<http://medicmobile.us8.list-manage.com/track/click?u=e5b07d2afb4b273eaff8e3c53&id=a3129fc799&e=c511575255>
.


Dianna Kane | Senior Designer
Medic Mobile <http://www.medicmobile.org/> +1 216 262 2204
@*M*edic <http://www.twitter.com/medic> Facebook
<https://www.facebook.com/MedicMobile> @diannakane

**Comment by garethbowen** [comment]
> When I get into select mode, the green dots shift to the left and the text shifts to the right, it looks weird

Fixed.

**Comment by garethbowen** [comment]
@estellecomment Is this ready for acceptance testing?

**Comment by estellecomment** [comment]
Sorry for the delay. Looking through the comments they're not all sorted out. You can file them separately if you want to move this along. 
 - [ ] rename to `DeleteDocs` (plural)
 - [ ] pass only `docs.length` to the delete confirm modal
 - [ ] in livelist, pass only a `isChecked` boolean to the insert/update/etc functions
The livelist refactoring I'll look into.
So almost done!

**Comment by garethbowen** [comment]
> rename to DeleteDocs

I'll do this soon.

> pass only docs.length to the delete confirm modal
> in livelist, pass only a isChecked boolean to the insert/update/etc functions

As per the comments I disagree with these.
